DataDog/dd-trace-rb
# Setup clangd-based native dev env via rake native_dev:setup

**What does this PR do?**
Add some helpers to facilitate development on the native extensions of `dd-trace-rb` using the [clangd language server](https://clangd.llvm.org/installation).

1. `rake native_dev:setup` task to generate the required `compile_commands.json` files through [bear](https://github.com/rizsotto/Bear) (following the [clangd getting started guide](https://clangd.llvm.org/installation#project-setup)).
2. `rake native_dev:clean` task to cleanup the files generated by the previous task (and integration into global `rake clean`).
3. Disable automatic clang formatting on the `ext/*` directories since the current source code format does not align with clang and it'll try to reformat whole files on every save.
4. Add a `.gitignore` specific to native extensions to ignore the `compilation-flags.json` files created from this setup and keep them out of source control.

**Motivation:**
Auto complete, navigation and diagnostic support during native extension development:

<img width="1502" alt="image" src="https://github.com/DataDog/dd-trace-rb/assets/373323/b89ee40d-e626-49f8-96ad-76f875481075">

<img width="1511" alt="image" src="https://github.com/DataDog/dd-trace-rb/assets/373323/5844f0a2-3396-41c0-8cba-8e5348f89261">

<img width="1507" alt="image" src="https://github.com/DataDog/dd-trace-rb/assets/373323/87279c93-a774-4404-87fb-0f6defad2932">


**Additional Notes:**


**How to test the change?**
1. Install [bear](https://github.com/rizsotto/Bear).
1. Run `rake native_dev:setup`.
2. Setup clangd LSP support in your IDE: [vscode](https://github.com/clangd/vscode-clangd) | [neovim](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#clangd) | [sublime](https://github.com/sublimelsp/LSP-clangd).
3. Check that completion and diagnostics work while editing files in `ext/`.

I tested it with `neovim` on my local setup.

**For Datadog employees:**
credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.

Unsure? Have a question? Request a review!
